### DISCUSSION TITLE
Define what the AST looks like

### DISCUSSION THREAD

**Discussion by adgad** [Financial-Times/content-tree#30] [General]
- We will create an extension of Unist
- We want to define nodes for each "tag", where the tags are given abstract (i.e. not specifically HTML) names - see https://github.com/Financial-Times/cp-content-pipeline/blob/main/packages/schema/src/tags/index.ts#L37

e.g. 
```
interface RecommendLink <: Parent {
  type: "element"
  recommendLinkId: string // these are taken from the XML attributes
  title: string
  children: [Element | Comment | Text] // TODO: do we want to define this strictly, or allow elements?
}
```

- We are **not** going to implement the `<experimental>` tag. Things that are currently being published inside that tag should have definitions.
   - _maybe_ if we need this kind of functionality, we should introduce a property. But lets not.
- For references, CP will still iterate through the AST to create the references array
   - This repo can provide Node definitions for "complete" nodes (i.e. a tweet with all teh embed stuff)
   - we can also provide helper functions to get the data (e.g. given a tweet ID, get the embed), which can be used in both CP and Spark Preview if required
   - somehow in our node definitions, differentiate between things which have an ID OR content

**Comment by chee** [comment]
https://github.com/Financial-Times/cp-content-pipeline/blob/main/packages/schema/src/tags/index.ts#L37-L74
```ts
  Link,
  Experimental: createTagWithoutReference('Experimental', 'experimental'),
  ImageSet,
  Recommended,
  RecommendedTitle: createTagWithoutReference(
    'RecommendedTitle',
    'recommended-title'
  ),
  Paragraph: createTagWithoutReference('Paragraph', 'p'),
  UnorderedList: createTagWithoutReference('UnorderedList', 'ul'),
  OrderedList: createTagWithoutReference('OrderedList', 'ol'),
  ListItem: createTagWithoutReference('ListItem', 'li'),
  Layout: createTagWithoutReference('Layout', '.n-content-layout, .layout'),
  LayoutContainer: createTagWithoutReference(
    'LayoutContainer',
    '.n-content-layout__container, .layout-container'
  ),
  LayoutSlot: createTagWithoutReference(
    'LayoutSlot',
    '.n-content-layout__slot, .layout-slot'
  ),
  LayoutImage,
  PullQuote,
  PullQuoteText: createTagWithoutReference('PullQuoteText', 'pull-quote-text'),
  PullQuoteSource: createTagWithoutReference(
    'PullQuoteText',
    'pull-quote-source'
  ),
  Blockquote: createTagWithoutReference('Blockquote', 'blockquote'),
  Tweet,
  LineBreak: createTagWithoutReference('LineBreak', 'br'),
  HorizontalRule: createTagWithoutReference('HorizontalRule', 'hr'),
  Emphasis: createTagWithoutReference('Emphasis', 'em'),
  Strong: createTagWithoutReference('Strong', 'strong'),
  Strikethrough: createTagWithoutReference('Strikethrough', 's'),
  Flourish,
  ```

this is the list of tags so far.

we'll also want `chapter`, `heading`, `subheading`... `scrollyblock`... not sure what else

**Comment by atanasdinov** [comment]
@chee you can use the following APIs to fetch the current internal and non-internal XSD schemas:
- `GET https://api.ft.com/schemas/body.xsd`
- `GET https://api.ft.com/schemas/body-internal.xsd`
